orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,157 @@ class PipelineSplitter {
   }
 };
 
+// TODO: move to a proper location for pipeline training
+
+// pipeline plan for each batch
+struct PipelineBatchInfo {
+  // Event pairs for each pipeline slot to WaitEvent when start, and RecordEvent when end
+  std::vector<std::pair<int64_t, int64_t>> events;
+  // indices of retired batches, so their data could be reused
+  // a batch can only be retired after finished backward in stage 0
+  // this can be used to join worker threads or reuse buffers
+  // for example, in a node with N GPUs and B batches to run in pipeline
+  // there will be (N * B) threads created, and by being able to retire,
+  // only (N * (2*N-1)) concurrent threads are needed.
+  // for big number of B, this would be helpful
+  std::vector<int64_t> retired_batches;
+};
+
+class PipelineTimeline {
+ public:
+  struct Slot {
+    enum class Type {
+      Forward,
+      Backward
+    };
+    Type type;
+    size_t batch_id;
+    bool occupied;
```

Comment:
Looks like this can be combined with the enum above if we consider `NotOccupied` as a new type? Every slot is initialized to `Empty` (or `NotOccupied`) and can be set to `Forward` or `Backward` afterwards.